File: 백준/Gold/9007. 카누 선수/카누 선수.cc
#include <bits/stdc++.h>
using namespace std;
typedef pair<int, int> pii;
typedef long long ll;
#define endl "\n"

struct Tree {
    int Node, left, right;
};

const int INF = 1e9;
const int MAX = 1001;
const int MOD = 1e9 + 7;
int dx[] = {0, 0, 1, -1};
int dy[] = {1, -1, 0, 0};
int T, k, n;

void solve(vector<int> &v1, vector<int> &v2) {
    vector<int> sum1, sum2;

    for(int i = 0; i < n; i++) {
        for(int j = n; j < 2 * n; j++) {
            sum1.push_back(v1[i] + v1[j]);
        }
    }

    for(int i = 0; i < n; i++) {
        for(int j = n; j < 2 * n; j++) {
            sum2.push_back(v2[i] + v2[j]);
        }
    }

    sort(sum1.begin(), sum1.end());
    sort(sum2.begin(), sum2.end());

    int answer = 0;
    int diff = INF;

    for(int i = 0; i < sum1.size(); i++) {
        int value = k - sum1[i];

        int idx = lower_bound(sum2.begin(), sum2.end(), value) - sum2.begin();

        if(idx < sum2.size()) {
            int cur_sum = sum1[i] + sum2[idx];
            int df = abs(k - cur_sum);

            if(cur_sum == k) {
                answer = cur_sum;
                break;
            }

            if(diff > df) {
                diff = df;
                answer = cur_sum;
            }
            else if(diff == df) {
                answer = min(answer, cur_sum);
            }
        }

        if(idx > 0) {
            int cur_sum = sum1[i] + sum2[idx-1];
            int df = abs(k - cur_sum);

            if(cur_sum == k) {
                answer = cur_sum;
                break;
            }

            if(diff > df) {
                diff = df;
                answer = cur_sum;
            }
            else if(diff == df) {
                answer = min(answer, cur_sum);
            }
        }
    }

    cout << answer << endl;
}

void input() {
    cin >> T;

    while(T--) {
        cin >> k >> n;

        vector<int> v1, v2;

        for(int i = 0; i < 2; i++) {
            for(int j = 0; j < 2 * n; j++) {
                int a;
                cin >> a;

                if(i == 0) v1.push_back(a);
                else v2.push_back(a);
            }
        }

        solve(v1, v2);
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);

    input();
}

File: 백준/Gold/9007. 카누 선수/카누 선수.java
import java.util.*;
import java.io.*;

public class Main {
    static int T, k, n;
    static int arr[][];

    static int lower_bound(int[] sum2, int goal) {
        int lo = -1;
        int hi = sum2.length;

        while(lo + 1 < hi) {
            int mid = (lo + hi) / 2;

            if(sum2[mid] >= goal) {
                hi = mid;
            }
            else lo = mid;
        }

        return hi;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        T = Integer.parseInt(br.readLine());

        while(T != 0) {
            T--;
            StringTokenizer st = new StringTokenizer(br.readLine());

            k = Integer.parseInt(st.nextToken());
            n = Integer.parseInt(st.nextToken());

            int[] sum1 = new int[n * n];
            int[] sum2 = new int[n * n];

            arr = new int[4][n];

            for(int i = 0; i < 4; i++) {
                st = new StringTokenizer(br.readLine());

                for(int j = 0; j < n; j++) {
                    arr[i][j] = Integer.parseInt(st.nextToken());
                }
            }

            int idx1 = 0;
            for(int i = 0; i < n; i++) {
                for(int j = 0; j < n; j++) {
                    sum1[idx1] = arr[0][i] + arr[1][j];
                    idx1++;
                }
            }

            idx1 = 0;
            for(int i = 0; i < n; i++) {
                for(int j = 0; j < n; j++) {
                    sum2[idx1] = arr[2][i] + arr[3][j];
                    idx1++;
                }
            }

            Arrays.sort(sum1);
            Arrays.sort(sum2);

            int answer = Integer.MAX_VALUE;
            int diff = Integer.MAX_VALUE;

            for(int i = 0; i < sum1.length; i++) {
                int value = k - sum1[i];

                int idx = lower_bound(sum2, value);

                if(idx < sum2.length) {
                    int val = sum1[i] + sum2[idx];

                    int df = Math.abs(k - val);

                    if(diff >= df) {
                        if(diff == df) {
                            answer = Math.min(answer, val);
                        }
                        else {
                            diff = df;
                            answer = val;
                        }
                    }
                }

                if(idx > 0) {
                    int val = sum1[i] + sum2[idx-1];

                    int df = Math.abs(k - val);

                    if(diff >= df) {
                        if(diff == df) {
                            answer = Math.min(answer, val);
                        }
                        else {
                            diff = df;
                            answer = val;
                        }
                    }
                }
            }

            bw.write(answer + "" + "\n");

            bw.flush();
        }
    }
}
